### DISCUSSION TITLE
empty tar

### DISCUSSION THREAD

**Discussion by uncertainty-principle** [NNPDF/eko#392] [Q&A] [answered]
hi.when i run the following code i get an empty myeko.tar and the program stops responding:
```py
import pathlib
import eko

from ekobox.cards import example

if __name__ ==  '__main__':
 th_card = example.theory()
 op_card = example.operator()
 op_card.xgrid = [1e-3, 1e-2, 1e-1, 5e-1, 1.]
 path = pathlib.Path("./myeko.tar")
 path.unlink(missing_ok=True)
 eko.solve(th_card, op_card, path)
```

and if i try to read the tar file using this code:

```py
import eko
with eko.EKO.read("./myeko.tar") as evolution_operator:
    # obtain theory card
    print(evolution_operator.theory_card)
    # or operator card
    print(evolution_operator.operator_card)
print(evolution_operator.evolgrid)
```

i get the following result:

> ```py
> TheoryCard(order=(1, 0), couplings=CouplingsInfo(alphas=0.118, alphaem=0.007496252, scale=91.2, max_num_flavs=6, num_flavs_ref=5, em_running=False), heavy=HeavyInfo(num_flavs_init=4, num_flavs_max_pdf=6, intrinsic_flavors=[4, 5, 6], masses=[[2.0, nan], [4.5, nan], [173.07, nan]], masses_scheme=<QuarkMassScheme.POLE: 'pole'>, matching_ratios=[1.0, 1.0, 1.0]), xif=1.0, n3lo_ad_variation=(0, 0, 0, 0, 0, 0, 0), use_fhmruvv=False, matching_order=(0, 0))
> OperatorCard(mu0=1.65, mugrid=[(100.0, 5)], xgrid=<eko.interpolation.XGrid object at 0x000001692FEBFAD0>, configs=Configs(evolution_method=<EvolutionMethod.ITERATE_EXACT: 'iterate-exact'>, ev_op_max_order=(10, 0), ev_op_iterations=10, scvar_method=None, inversion_method=None, interpolation_polynomial_degree=4, interpolation_is_log=True, polarized=False, time_like=False, n_integration_cores=0), debug=Debug(skip_singlet=False, skip_non_singlet=False), eko_version='0.14.3')
> []
> ```

**Comment by felixhekhorn** [comment, marked as answer]
Hey @uncertainty-principle ,
thanks for reaching out.

Unfortunately I can not reproduce your behavior - instead for me it works as expected and I get the expected tuple `[(10000.0, 5)]` back for the last print.

However, I'm guessing the problem is here: "the program stops responding" - unfortunately this is currently a "bad feature" of eko. The problem is eko is relying on [numba](https://numba.pydata.org/) for a fast evaluation of the underlying integrals. This implies a compilation from Python to machine code (which is then quick to execute) and this compilation is done just-in-time (JIT), which in practice is on the first actual call. Since we have a quite big code base (with many mathematical non-trivial functions) this actually requires some time and it may indeed seem that the program stops responding. Depending on the complexity you ask (and your machine) this can take up to 30 min. However, once the compilation is done it is cached and then reused every time you ask eko again.

Can you please try again and let it run a bit longer? I would also recommend to adjust the operator card for the first run (and as explained only the first) and set `op_card.configs.n_integration_cores = 1` (which will thus not use parallel cores).

We are working on a solution to the problem by using Rust for the hard math and thus compiling at wheel creation time.

PS: Note that the `print(evolution_operator.evolgrid)` should be scoped under the `with`-statement to have the variable `evolution_operator` properly in scope ("by chance" it also works like this). However, this might have been introduced unintentionally by @alecandido when re-formatting your post to proper code blocks ...

**Reply by alecandido** [reply]
> However, this might have been introduced unintentionally by @alecandido when re-formatting your post to proper code blocks ...

Not really, you can check the original comment. It seemed weird, but I decided to leave it as it was.

**Reply by uncertainty-principle** [reply]
You were right! the problem was i tried to run EKO while running another resource-intensive program(using Matlab) which made the system unresponsive. Once i allocated more resources (and time) to EKO,it produced the expected results!
I did not change any settings though(i.e setting op_card.configs.n_integration_cores = 1) as i decided to to that as a last resort.
Thanks a million for your help!

**Reply by felixhekhorn** [reply]
I'm glad to hear the problem was solved quickly :+1: Feel free to contact us again if you have further questions.
